crypto/s2n_certificate.c

```diff
@@ -47,45 +47,30 @@ int s2n_create_cert_chain_from_stuffer(struct s2n_cert_chain *cert_chain_out, st
 
     struct s2n_cert **insert = &cert_chain_out->head;
     uint32_t chain_size = 0;
-    do {
-        struct s2n_cert *new_node = NULL;
+    while (s2n_stuffer_pem_has_certificate(chain_in_stuffer)) {
+        int result = s2n_stuffer_certificate_from_pem(chain_in_stuffer, &cert_out_stuffer);
+        POSIX_ENSURE(result == S2N_SUCCESS, S2N_ERR_INVALID_PEM);
 
-        if (s2n_stuffer_certificate_from_pem(chain_in_stuffer, &cert_out_stuffer) < 0) {
-            if (chain_size == 0) {
-                POSIX_BAIL(S2N_ERR_NO_CERTIFICATE_IN_PEM);
-            }
-            break;
-        }
-        struct s2n_blob mem = { 0 };
+        DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
         POSIX_GUARD(s2n_alloc(&mem, sizeof(struct s2n_cert)));
         POSIX_GUARD(s2n_blob_zero(&mem));
-        new_node = (struct s2n_cert *) (void *) mem.data;
 
-        if (s2n_alloc(&new_node->raw, s2n_stuffer_data_available(&cert_out_stuffer)) != S2N_SUCCESS) {
-            POSIX_GUARD(s2n_free(&mem));
-            S2N_ERROR_PRESERVE_ERRNO();
-        }
-        if (s2n_stuffer_read(&cert_out_stuffer, &new_node->raw) != S2N_SUCCESS) {
-            POSIX_GUARD(s2n_free(&mem));
-            S2N_ERROR_PRESERVE_ERRNO();
-        }
+        struct s2n_cert *new_node = (struct s2n_cert *) (void *) mem.data;
+        POSIX_GUARD(s2n_alloc(&new_node->raw, s2n_stuffer_data_available(&cert_out_stuffer)));
+        POSIX_GUARD(s2n_stuffer_read(&cert_out_stuffer, &new_node->raw));
+        ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
 
         /* Additional 3 bytes for the length field in the protocol */
         chain_size += new_node->raw.size + 3;
         new_node->next = NULL;
         *insert = new_node;
         insert = &new_node->next;
-    } while (s2n_stuffer_data_available(chain_in_stuffer));
-
-    /* Leftover data at this point means one of two things:
-     * A bug in s2n's PEM parsing OR a malformed PEM in the user's chain.
-     * Be conservative and fail instead of using a partial chain.
-     */
-    S2N_ERROR_IF(s2n_stuffer_data_available(chain_in_stuffer) > 0, S2N_ERR_INVALID_PEM);
```

Comment:
> If the characters aren't present, then there are no more certificates in the pem file (although there might be whitespace / comments / other garbage, but we also skipped over those before).

How did we skip over arbitrary trailing content before? Doesn't this line check for this?

And maybe it's fine to permit arbitrary trailing content since we do that between certificates anyway? But just wanted to check to see if this is a behavior change, and if so that it was intended.

Reply:
First of all, definitely not a behavior change. I added tests for this, and they still pass with my fix reverted.

But why did it work before? Oh man I'm so glad you asked. This is a fun one.

As long as there is data available in the pem, we will continue to attempt to read certificates from it. The first step of reading a certificate is [to read until the "--"](https://github.com/aws/s2n-tls/blame/d9df7849d4dc5356e3b54e15a8e0d2e273327265/stuffer/s2n_stuffer_pem.c#L44). So when there's gibberish at the end of the chain, we read all the gibberish, then fail because [we can't read  the "--"](https://github.com/aws/s2n-tls/blob/d9df7849d4dc5356e3b54e15a8e0d2e273327265/stuffer/s2n_stuffer_text.c#L92). So when we break out of the loop, there's no gibberish left for that check to detect. It's not the most readable code, but it does work.

If you're wondering, it also worked before my previous fix to parse comments with '-' in them. Back then the first step was [to read until the '-'](https://github.com/aws/s2n-tls/blame/cd9eaa5fd589e7e8e785203c7dc4792e25f56436/stuffer/s2n_stuffer_pem.c#L40). We then failed when we [tried to read at least one '-'](https://github.com/aws/s2n-tls/blame/cd9eaa5fd589e7e8e785203c7dc4792e25f56436/stuffer/s2n_stuffer_pem.c#L43).

Reply:
I guess I added a test for final comments and final whitespace, but not for final gibberish that doesn't start with a "#". The distinction doesn't actually matter, but maybe a true "random gibberish" test would make that behavior clearer.

Reply:
Ohhh right ok, that makes sense!